### PR TITLE
Allow authentication context requests

### PR DESCRIPTION
**Pull request by avdata99**
Related to [#24](https://github.com/keitaroinc/ckanext-saml2auth/issues/24) and [GSA#2526](https://github.com/GSA/datagov-deploy/issues/2526)

This PR:
 - Add new setting value to allow authentication context
 - Remove Travis and start GitHub action for CI

In the process of login or if I create new users using the `ckanext.saml2auth.requested_authn_context` setting we can see PIV cards (or other methods) are allowed

With `http://idmanagement.gov/ns/assurance/aal/3?hspd12=true`

Selection_014

With `http://idmanagement.gov/ns/assurance/aal/3` and (probably not required both) `http://idmanagement.gov/ns/assurance/aal/3?hspd12=true`

Selection_017

### CODE REVIEW

File: ckanext/saml2auth/views/saml2auth.py [hunk 1]
```diff
@@ -128,12 +130,34 @@ def acs():
     return resp
 
 
+def get_requested_authn_contexts():
+    requested_authn_contexts = config.get('ckanext.saml2auth.requested_authn_context', None)
+    if requested_authn_contexts is None:
+        return []
+
+    return requested_authn_contexts.strip().split()
```
Comment (adborden) [inline ckanext/saml2auth/views/saml2auth.py:138]: I forget how `split()` works, do you need a compact() to remove empty strings in the case of a double space between two reqs?

**Review by adborden — APPROVED** [review]

File: ckanext/saml2auth/views/saml2auth.py (hunk 1 above)
Comment (avdata99) [reply ckanext/saml2auth/views/saml2auth.py:138]:
```
a = "one two  three      four"
a.strip().split()
['one', 'two', 'three', 'four']
```
I think it's ok

**Comment by mbocevski** [comment]
If this would be part of the extension as proposed by this implementation, then I think we should also support `Comparison` as part of the request authn context, cause it's very commonly used especially when multiple contexts are being requested.

**Comment by avdata99** [comment]
@mbocevski, as you said, `comparison` is important
I [added this feature](https://github.com/keitaroinc/ckanext-saml2auth/pull/26/commits/023b9f8b4c8444f08ca8f09b89cae57357535e89) and a [simple test](https://github.com/keitaroinc/ckanext-saml2auth/pull/26/commits/1cf265a292edcfc6445182cb7f47bdfef0c2d56b) for it

I think is enough. Let me know
cc: @adborden

**Comment by mbocevski** [comment]
@avdata99 yeah looks good. Just take care of the rest of the review comments and I think this is ready to merge!

**Comment by avdata99** [comment]
@mbocevski I think there are no pending review comments. Maybe i'm missing something
Maybe @duskobogdanovski want to add something. Let me know.

**Review by mbocevski — CHANGES_REQUESTED** [review]

**Comment by mbocevski** [comment]
@avdata99 so sorry, I have forgotten to submit the review. Can you please go through the comments.

**Comment by duskobogdanovski** [comment]
Except for @mbocevski 's comments the rest looks good to me.

File: .github/workflows/checks.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
+name: CI
+
+on:
+  push:
+    branches:
+      - '**'
```
[the remark this answers is not in the document]
Comment (avdata99) [reply .github/workflows/checks.yml]: Done

File: .github/workflows/checks.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      - name: Install nosetests
+        run: |
+          python -m pip install pip==20.3.3
+          pip install nose
+
+      - name: Setup CKAN
+        env:
+          PGPASSWORD: postgres
+        run: |
+          bash bin/travis-build.bash
```
[the remark this answers is not in the document]
Comment (avdata99) [reply .github/workflows/checks.yml]: Done

File: dev-requirements.txt [hunk 4]
```diff
@@ -1,2 +1,3 @@
 flake8    # for the travis build
```
[the remark this answers is not in the document]
Comment (avdata99) [reply dev-requirements.txt]: Done

File: ckanext/saml2auth/views/saml2auth.py [hunk 5]
```diff
@@ -143,11 +143,14 @@ def saml2login():
      configured identity provider for authentication
     '''
     client = h.saml_client(sp_config())
-    requested_authn_context = None
     requested_authn_contexts = get_requested_authn_contexts()
 
     if len(requested_authn_contexts) > 0:
-        requested_authn_context = RequestedAuthnContext()
+        comparison = config.get('ckanext.saml2auth.requested_authn_context_comparison', 'exact')
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/saml2auth/views/saml2auth.py]: Done

File: ckanext/saml2auth/views/saml2auth.py [hunk 6]
```diff
@@ -128,12 +130,34 @@ def acs():
...
+    requested_authn_contexts = get_requested_authn_contexts()
+
+    if len(requested_authn_contexts) > 0:
+        requested_authn_context = RequestedAuthnContext()
+
+        for item in requested_authn_contexts:
+            context_class_ref = AuthnContextClassRef()
+            context_class_ref.text = item
+            log.debug('requested_authn_context added {}'.format(item))
+            requested_authn_context.authn_context_class_ref.append(context_class_ref)
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/saml2auth/views/saml2auth.py]: I now use that pysaml2 function which seems clear.

**Review by duskobogdanovski — APPROVED** [review]

**Pull request merged**